controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,588 @@
+	if err != nil {
+		return errors.Wrapf(err, "failed to get hash status for snapshot %v", snapshotName)
+	}
+
+	for _, status := range hashStatus {
+		if status.State == string(SnapshotHashStatusError) {
+			return fmt.Errorf("failed to hash snapshot %v since %v", snapshotName, status.Error)
+		}
+
+		if status.State == string(engineapi.ProcessStateInProgress) {
+			return fmt.Errorf(string(engineapi.ProcessStateInProgress))
+		}
+	}
+
+	snapshot, err := m.ds.GetSnapshot(snapshotName)
+	if err != nil {
+		return err
+	}
+	existingSnapshot := snapshot.DeepCopy()
+
+	checksum, err := determineChecksumFromHashStatus(snapshotName, snapshot.Status.Checksum, hashStatus)
+	if err != nil {
+		m.eventRecorder.Eventf(engine, v1.EventTypeWarning, constant.EventReasonFailedSnapshotDataIntegrityCheck,
+			"Failed to check the data integrity of snapshot %v for volume %v", snapshotName, engine.Spec.VolumeName)
+		return errors.Wrapf(err, "failed to determine checksum for snapshot %v", snapshotName)
+	}
+
+	snapshot.Status.Checksum = checksum
+
+	if !reflect.DeepEqual(existingSnapshot.Status, snapshot.Status) {
+		if _, err := m.ds.UpdateSnapshotStatus(snapshot); err != nil {
+			return errors.Wrapf(err, "failed to update status for snapshot %v", snapshotName)
+		}
+	}
+
+	m.kickOutCorruptedReplicas(engine, engineClientProxy, checksum, hashStatus)
+
+	return nil
+}
+
+func (m *SnapshotMonitor) kickOutCorruptedReplicas(engine *longhorn.Engine, engineClientProxy engineapi.EngineClientProxy,
+	checksum string, hashStatus map[string]*longhorn.HashStatus) {
+	for address, status := range hashStatus {
+		if status.Checksum == checksum {
+			continue
+		}
+
+		m.eventRecorder.Eventf(engine, v1.EventTypeWarning, constant.EventReasonFaulted, "Detected corrupted replica %v", address)
+
+		if err := engineClientProxy.ReplicaModeUpdate(engine, address, string(etypes.ERR)); err != nil {
+			m.logger.WithField("monitor", monitorName).Errorf("failed to update replica %v mode to ERR", address)
+		}
+	}
+}
+
+func determineChecksumFromHashStatus(snapshotName, existingChecksum string, hashStatus map[string]*longhorn.HashStatus) (string, error) {
+	checksums := map[string][]string{}
+
+	for address, status := range hashStatus {
+		// The checksum from a silently corrupted snapshot disk file cannot vote.
+		if status.SilentlyCorrupted {
+			continue
+		}
+		checksums[status.Checksum] = append(checksums[status.Checksum], address)
+		logrus.Debugf("Replica %v snapshot %v checksum %v", address, snapshotName, status.Checksum)
+	}
+
+	if len(checksums) == 0 {
+		return "", fmt.Errorf(prefixChecksumDetermineFailure+"since snapshot disk files are silently corrupted", snapshotName)
+	}
+
+	found, checksum, maxVotes := determineChecksum(checksums)
+	if found {
+		return checksum, nil
+	}
+
+	if existingChecksum == "" {
+		return "", fmt.Errorf(prefixChecksumDetermineFailure+"since there is no existing checksum", snapshotName)
+	}
```

Comment:
Should we error out here? For a newly hashed snapshot, this should be empty... How about directly input `existingChecksum` as an input of `determineChecksum`? `determineChecksum` will use the vote mechanism only if `existingChecksum` is empty.

Reply:
> Should we error out here? For a newly hashed snapshot, this should be empty...

For the case of a newly hashed snapshot, the checksum should be determined by the previous
```
	found, checksum, maxVotes := determineChecksum(checksums)
	if found {
		return checksum, nil
	}
```

> How about directly input existingChecksum as an input of determineChecksum? determineChecksum will use the vote mechanism only if existingChecksum is empty.

The checksums from replicas might be different than prevous values, because purge, trim, corruption and etc. So, the vote mechanism is alway executed to get the latest checksum and then update the `status.checksum`.

Reply:
OK. Got it. Please add a comment based on this `The checksums from replicas might be different than prevous values, because purge, trim, corruption and etc`

Besides, can this monitor print out a log if there is a checksum change?

Reply:
Sure. Will add the log.